Blogpost about Snapchat Minis (#444)
ombulabs/blog: _posts/2020-07-27-creating-a-snap-chat-mini.markdown

---
layout: post
title: "Crossbrowser Mobile App - What we Learned Developing a Snapchat Mini"
date: 2020-07-27 12:00:00
categories: ["learning"]
author: arieljuod
---

Over the last few months, we were making a Snapchat Mini <!-- I'll link to Amanda's post -->. Minis are small static web apps that are run inside a webview within the Snapchat native app. One important part of the development was to make sure our minis works well across different devices, specially making sure things work the same in both Android's webview (Chrome by default) and iOS's webview (Safari).

In this post I'll comment a few considerations when working with the Span canvas SDK and also some common and known issues you'll encounter when making your app cross-mobile-browser compatible.

<!--more-->

## Working with the SDK

> Snapchat provides a JavaScript SDK that you'll have to include in your project. This SDK works as a bridge between the native app and your web app and lets you trigger native functionalities (like sharing on chat and camera), it gives you information of the current user and session, and also emits different events so you can react to them (you can detect if a user took a screenshot or changed the volume for example).

Author: I'll replace `emits` with `sends` since that's the word used in snapchat's manual.

```javascript
let base64data = generateStickerData()
if (sc.app) {
  // this method will prepare the sticker object and call the SDK methods
  shareStickerUsingSDK(base64Data) 
} else {
  // this will show an IMG element during development locally, so we can see what we created
  previewImage.src = base64data
}
```

This speeds up the sticker design process a lot! since we don't need to publish the mini to test it inside Snapchat. After we were happy with the generated image, we started testing the real sdk method calls publishing the mini to be used inside Snapchat.

## Embrace the Asynchronous Nature of JavaScript

If you are comming from other programming languages, you may find hard to work with callbacks (you may end up falling into the [Callback Hell](http://callbackhell.com)). Instead of trying to adapt the SDK methods to a synchronous nature, we found it's a lot simpler to use modern JavaScript techniques to handle asynchronism. We tried to reduce callbacks using Promises and we also used reactive JavaScript frameworks like VueJS or ReactJS so we could update the app's state asynchronously and let those fraemworks update the views.

## Storing Data

> Depending on the nature of the mini, you may need to store some data related to the user's state/actions (you may want to store an ID you get from a service, a timestamp when a user did some action, etc). We needed to store if a user had already done one specific action. For simple data like this, you can use the SDK's provided methods to write and read the local storage. Don't confuse this with the brower's local storage API: the local storage provided by the SDK is similar in some aspects but (apart from a few small differences) it has an async nature while the browser's locale storage is synchronous.

Author: What about `In our case, we needed to store ...`?

You can use it to store simple string key/value pairs. You can store small and simple object by stringifying them first, but you shouldn't abouse this.

If you need to store more data, more complex data, or if you need a database to share data between users, you'll need to handle that within your own infraestructure.

## Native Look and Feel

> We wanted this mini too look and feel like it was a native app. Some key considerations were: the design should not look like a web, sizes, color and animations/transitions should feel similar to what native apps do, and we wanted a scrolling behaviour that didn't feel like a web.

Author: I meant a webapp/website, yes.

For the scrolling, we implemented a custom scroll on one of the apps because it relays too much in swiping and scrolling gestures and the default scroll was not enough for our needs. This was developed with this mini's requirements in mind so it's something that probably won't work for other minis to share what we did, but the main idea is to listen to the touch events (start, move, stop, cancel) and use its position to detect the gestures and scroll the elements accordingly. We also added some inertia like it happens with the default scroll when releasing the touch.

You can try to add support for similar mouse events (desktop Safari mobile simulation mode does not trigger touch events like Chrome does for example), but minis are only going to be run inside touch devices. It could help you during development when testing inside desktop Safari if you need that.

## Performance

One of the apps was animation-intensive: colors changing, text scrolling, different parts could move at the same time, some actions would trigger big changes on the layout, etc.

> It was really important to measure and test the performance. The most important tool we used for this was Google Chrome's `Performance` profiler in dev tools. You can profile the app while doing specific actions and then inspect it frame by frame to make sure what you need to optimize to have smoth animations.

Author: I'll leave the `optimize` word here, using the profiler you are not of sure what to do, you can see what actions are consuming time (maybe it takes too much time executing javascript, maybe too much time recalculating layout, maybe repainting) but it's up to you then try different things to make it smoother.

My advice here is that, if you have a form, you should have an specific `submit` element. If you relay on the user pressing `enter` or `return` keys, that might be counterintuitive for iOS users used to press `DONE` when they finished typing.

### Repaint Issues

When you are working with animations, the browsers will try to optimize the repaints so improve performance. We encountered some really specific cases where, for example, an animation happened too quick outside the viewport and it had the wrong final state.

The solution we had here was to force a repaint of the element at the time we detected it was inside the viewport:

```javascript
afterEnteringTheViewport(element) {
  element.style.transform = 'translateZ(0)'
}
```

> This is just an example of a function, it's not a real callback function you can use like an event handler. How you detect that the element enters the viewport will depend on your app's implementation. You could use the `IntersectionObserver` for example or some css animation-related events.

> ### Safari can't Handle `box-shadow` and a Gradient Background

Reviewer: Hmmm, I'm not sure that this is specific to snap minis and therefor if it belongs here?

If you need a gradient background AND a box shadow in the same element, you may have to create an auxiliary element (or a `:before` pseudo element) absolute positioned behind all the other elements to serve as background. Then you can make that background element have the gradient, and the original element have the box-shadow.

### The Notch

iOS introduced `The Notch` a few years ago. This hardware element will cover a part of the screen. Your mini will be displayed behind the notch, so you'll have to take that into consideration when designing your app, so you don't put elements fixed at the top that might be obscured by the notch.

You can use CSS' safe area inset values to add extra padding when needed.

This is another great article from css-tricks you can use for more in-depth code an examples: [link](https://css-tricks.com/the-notch-and-css/).

If you need to know these values in your JavaScript code, the SDK provides that information (using `sc.app.safeAreaInsets`) with the top and bottom safe area inset values.

### iOS Overscroll

> Apple users are used to the extra scroll behavior. When the user reaches the end of something that's scrollable, they can scroll a bit more, showing up a white space on the direction they are over scrolling.

Reviewer: ```suggestion
Apple users are used to the extra scroll behavior. When the user reaches the end of something that's scrollable, they can scroll a bit more. A white space shows up in the direction they over-scrolled.
```

> If you want to prevent that overscroll to shift your whole app outside the viewport, you can position your app fixed to the viewport, so if the user overscrolls it, it will scroll the static element in the background but your app will be fixed on top so the scroll will not affect it. It may be hard to explain, and the code for this is just a `position: fixed` on an element, hopefully this explanation is enough if you are experiencing the same issue.

Reviewer: ```suggestion
If you want to prevent that overscroll to shift your whole app outside the viewport, you can position your app fixed to the viewport, so if the user overscrolls it, it will scroll to the static element in the background but your app will be fixed on top so the scroll will not affect it. It may be hard to explain, and the code for this is just a `position: fixed` on an element, hopefully this explanation is enough if you are experiencing the same issue.
```

> ### Android Keyboar Covers Screen

Reviewer: ```suggestion
### Android Keyboard Covers Screen
```

> Android and iOS devices does different things when you open the keyboard. Each of them has pros and cons. One key difference is that when you open the keyboard on iOS, it will try to scroll so the input is inside the viewport and if it can scroll to it it will move the whole app outside the screen if necessary so the input is visible. For Android, it will only try to scroll but if the element is to look inside the viewport, it will be covered by the keyboard. You may need to take this into consideration when designing.

Reviewer: ```suggestion
Android and iOS devices do different things when you open the keyboard. Each of them has pros and cons. One key difference is that when you open the keyboard on iOS, it will try to scroll so the input is inside the viewport and if it can scroll to it it will move the whole app outside the screen if necessary so the input is visible. For Android, it will only try to scroll but if the element is to look inside the viewport, it will be covered by the keyboard. You may need to take this into consideration when designing.
```
The android part of this is a little confusing. Consider reworking it.

## Conclusion

> We were able to fix or workaround a lot of inconsistencies between Chrome for Android and iOS' Safari, and we wanted to share the issues we had and the solutions we've found so far. Some solutions may not have a clear code snippet to apply them, but I hope the explanations are clear enough so you can implement your own solutions when working with minis or mobile webapps.

Reviewer: ```suggestion
We were able to fix or workaround a lot of inconsistencies between Chrome for Android and iOS' Safari, and we wanted to share the issues we had and the solutions we've found so far. Some solutions may not have a clear code snippet to apply to them, but I hope the explanations are clear enough so you can implement your own solutions when working with minis or mobile webapps.
```

---
layout: post

> title: "What we Learned Developing a Snapchat Mini (Part 1)"

Reviewer 2: @arielj I vote for updating the title to `"The Chronicles of the Minis Master"` :squirrel:
